src/thermo/Phase.cpp

```diff
@@ -256,42 +256,122 @@ std::string Phase::speciesSPName(int k) const
     return m_name + ":" + speciesName(k);
 }
 
+vector<std::string> Phase::defaultState() const
+{
+    if (isPure()) {
+        if (isCompressible()) {
+            return {"T", "density"};
+        } else {
+            return {"T", "P"};
+        }
+    } else {
+        if (isCompressible()) {
+            return {"T", "density", "Y"};
+        } else {
+            return {"T", "P", "Y"};
+        }
+    }
+}
+
+vector<std::string> Phase::fullStates() const
+{
+    if (isPure()) {
+        if (isCompressible()) {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        } else {
+            return {"TP", "HP", "SP"};
+        }
+    } else {
+        if (isCompressible()) {
+            return {"TDX", "TDY", "TPX", "TPY", "UVX", "UVY", "DPX", "DPY",
+                    "HPX", "HPY", "SPX", "SPY", "SVX", "SVY"};
+        } else {
+            return {"TPX", "TPY", "HPX", "HPY", "SPX", "SPY"};
+        }
+    }
+}
+
+vector<std::string> Phase::partialStates() const
+{
+    if (isPure()) {
+        return {};
+    } else {
+        if (!isCompressible()) {
+            return {"TP", "HP", "SP"};
+        } else {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        }
+    }
+}
+
 void Phase::saveState(vector_fp& state) const
 {
-    state.resize(nSpecies() + 2);
+    if (isPure()) {
+        state.resize(2);
+    } else {
+        state.resize(nSpecies() +2);
+    }
     saveState(state.size(), &state[0]);
 }
+
 void Phase::saveState(size_t lenstate, doublereal* state) const
 {
     state[0] = temperature();
-    state[1] = density();
-    getMassFractions(state + 2);
+    if (isCompressible()) {
+        state[1] = density();
+    } else {
+        state[1] = pressure();
+    }
+    if (!isPure()) {
+        if (defaultState()[2] == "Y") {
```

Comment:
I'm not a big fan of this "magic number" approach to finding the mass or mole fraction default. There is an implicit requirement that overriding functions keep the same order (i.e., mole or mass fraction last). However, I recognize that there is value in the simplicity of having just `std::vector<std::string>` as the type. Is there any way to remove this magic number and keep the simplicity (he asked without much hope)? If not, can this requirement at least be documented?

Reply:
I'm not sure whether this is worth the extra complexity, but `defaultState()` could return a `std::map<std::string, size_t>` where the keys are the property names and the values are the offsets into the state vector. It wouldn't be a bad structure for the (future) case where one wants to include additional state variables.

Reply:
That's a valid point. I followed @speth's idea to implement a map here, which eliminates the magic numbers and allows for future additional states.